### ISSUE TITLE
Generating a schema from a type with a nullable object causes an error. (v3.7.0)

### ISSUE THREAD

**Issue by yumemi-tokumoto** [fabien0102/ts-to-zod#203]
## Bug description

Error occurs and schema generation fails.
Generating a schema from a type with a nullable object causes an error. (v3.7.0)

### Input

```ts
export type User = {
  name: string
}

export type Article = {
  user: User | null
}
```

### Expected output

```ts
export const userSchema = z.object({
  name: z.string(),
})

export const articleSchema = z.object({
  user: userSchema.nullable(),
})
```

### Actual output

```
✖ Validating generated types
 ›   Error: 'Article' is not compatible with 'articleSchema':
 ›   Argument of type 'Article' is not assignable to parameter of type '{ user: { name: string; } | null; }'.
 ›     Types of property 'user' are incompatible.
 ›       Type 'User | null | undefined' is not assignable to type '{ name: string; } | null'.
 ›         Type 'undefined' is not assignable to type '{ name: string; } | null'.
```

**Versions**

- Typescript: `v5.3.3`
- Zod: `v3.22.4`

**Comment by yumemi-tokumoto** [comment]
Until I updated to version 3.6.1, I was able to generate it without error.

**Comment by schiller-manuel** [comment]
@tvillaren can you have a look please? seems like your latest changes introduced a regression.

**Comment by adamwdennis** [comment]
I am also getting this error.

**Comment by tvillaren** [comment]
> @tvillaren can you have a look please? seems like your latest changes introduced a regression.

I'll have a look

**Comment by schiller-manuel** [comment]
This can be reproduced by running 

```
$ yarn gen:example
```

output:
```
yarn run v1.22.19
warning ../../package.json: No license field
$ ./bin/run --config example
- Validating generated types
✖ Validating generated types
 ›   Error: 'SkillsSpeedEnemy' is not compatible with 'skillsSpeedEnemySchema':
 ›   Argument of type 'SkillsSpeedEnemy' is not assignable to parameter of type
 ›    '{ power: EnemyPower.Speed; }'.
 ›     Types of property 'power' are incompatible.
 ›       Type 'EnemyPower.Speed | undefined' is not assignable to type 
 ›   'EnemyPower.Speed'.
 ›         Type 'undefined' is not assignable to type 'EnemyPower.Speed'.
 ›   'Superman' is not compatible with 'supermanSchema':
 ›   Argument of type 'Superman' is not assignable to parameter of type '{ 
 ›   name: "superman" | "clark kent" | "kal-l"; enemies: Record<string, any>; 
 ›   age: number; powers: ["fly", "laser", "invincible"]; underKryptonite?: 
 ›   boolean | undefined; }'.
 ›     Types of property 'enemies' are incompatible.
 ›       Type 'Record<string, Enemy> | undefined' is not assignable to type 
 ›   'Record<string, any>'.
 ›         Type 'undefined' is not assignable to type 'Record<string, any>'.
 ›   'Exported' is not compatible with 'exportedSchema':
 ›   Argument of type 'Exported' is not assignable to parameter of type '{ a: {
 ›    name: string; }; b: string; }'.
 ›     Types of property 'a' are incompatible.
 ›       Type 'NonExported | undefined' is not assignable to type '{ name: 
 ›   string; }'.
 ›         Type 'undefined' is not assignable to type '{ name: string; }'.
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

I see 3 separate issues:
1. most obvious: why didn't our tests find this?
2. `| undefined` is somehow added to referenced types in the validation
3. if a type is imported from a separate file, it correctly generates the type but during validation this separate file is not imported; see `Record<string, any>` above which should be `Record<string, Enemy>`;

@tvillaren do you typically use `--skipValidation`?

**Comment by tvillaren** [comment]
Yes, the use case is specific to `TypeReference` being set as optional by `fixOptionalAny` during the validation step. 

This was introduced for `any` keyword due a known issue with `ts-to-zod` (see https://github.com/fabien0102/ts-to-zod/issues/140) but I updated this method to handle external imports (which are inferred as `any` as well in our validation step). 

Pushed a failing test to reproduce: https://github.com/fabien0102/ts-to-zod/compare/main...tvillaren:ts-to-zod:fix-issue-203
The issue is that we now mark all TypeReference as optional while we should do so only for imported ones...

**Comment by mpiltz** [comment]
We have run in to this same issues when we are generating TS Type Definitions from OpenAPI Schemas and then genarate zod-schemas from the generated TS types and interfaces. 

TL;DR; zod-schema generation fails with same error from this:


`
export type MyField = string
export interface MyObject {
 myField: MyField
}

`

**Comment by johnnyp-gg** [comment]
Im having the same issue with:
```
Input
export type User = {
  username: string | undefined
}
```
```
Expected output
export const userSchema = z.object({
  name: z.string().optional(),
})
```
on version 3.8.3 could it need the same fix?

**Comment by tvillaren** [comment]
@johnnyp-gg: actually

```ts
export type User = {
  username: string | undefined
}
```

generates (with `--skipValidation` flag)

```ts
// Generated by ts-to-zod
import { z } from "zod";

export const userSchema = z.object({
  username: z.union([z.string(), z.undefined()]),
});
```

which is the expected schema. 
Indeed, it's not the same to check for an optional field and for a mandatory one which may take the `undefined` value. 
The validation error you're getting comes from a known bug in Zod inference (https://github.com/colinhacks/zod/issues/2464)
